### PR TITLE
ISPN-15974 Protocol buffers schema compatibility checks

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-15974

Depends on https://github.com/infinispan/protostream/pull/260

### CODE REVIEW

File: core/src/main/resources/schema/infinispan-config-15.0.xsd [hunk 1]
```diff
@@ -3323,4 +3330,32 @@
     </xs:restriction>
   </xs:simpleType>
 
+  <xs:simpleType name="schema-compatibility">
+    <xs:restriction base="xs:token">
+      <xs:enumeration value="UNRESTRICTED">
+        <xs:annotation>
+          <xs:documentation>
+            Replaces any existing schemas without performing any backwards-compatibility checks.
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply core/src/main/resources/schema/infinispan-config-15.0.xsd:3338]: Done

File: core/src/main/java/org/infinispan/configuration/global/SerializationConfiguration.java [hunk 2]
```diff
@@ -26,9 +27,11 @@ public class SerializationConfiguration {
    public static final AttributeDefinition<List<SerializationContextInitializer>> SERIALIZATION_CONTEXT_INITIALIZERS =
          AttributeDefinition.builder("contextInitializers", null, (Class<List<SerializationContextInitializer>>) (Class<?>) List.class)
                .immutable().build();
+   public static final AttributeDefinition<Configuration.SchemaValidation> SCHEMA_COMPATIBILITY_VALIDATION =
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/global/SerializationConfiguration.java]: Done

**Comment by ryanemerson** [comment]
@fax4ever Can you also take a look as this impacts Query?

**Comment by fax4ever** [comment]
> @fax4ever Can you also take a look as this impacts Query?

Hi @ryanemerson. Sure, it sounds a good idea

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/evolution/model/BaseModelWithNameAnalyzedAndNameNonAnalyzedFieldEntity.java [hunk 3]
```diff
@@ -22,19 +22,19 @@ public class BaseModelWithNameAnalyzedAndNameNonAnalyzedFieldEntity implements M
...
     @Deprecated(forRemoval=true)
-    @Text()
-    public String nameAnalyzed;
+    @Text
+    public String name;
 
     @ProtoField(number = 4)
     @Basic(projectable = true)
-    public String nameNonAnalyzed;
+    public String analyzed;
```
Comment (fax4ever) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/evolution/model/BaseModelWithNameAnalyzedAndNameNonAnalyzedFieldEntity.java:30]: I think it is not analyzed

File: server/tests/src/test/resources/proto/v1.proto [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+message m1 {
+  string f1 = 1;
+  int32 f2 = 2;
+  bool f3 = 3;
+  map<string,string> f4 = 4;
+
+  reserved 5, 6, 7;
+  reserved "f5", "f6";
+}
```
Comment (fax4ever) [inline server/tests/src/test/resources/proto/v1.proto:22]: Is this file used by some test? (also the other in the same folder)

**Review by fax4ever — APPROVED** [review]
Great contribution. I really like the idea of having `strict` as a default. Just 2 very-minor questions...

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/evolution/model/BaseModelWithNameAnalyzedAndNameNonAnalyzedFieldEntity.java (hunk 3 above)
Comment (tristantarrant) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/evolution/model/BaseModelWithNameAnalyzedAndNameNonAnalyzedFieldEntity.java:30]: I know, but I was concerned mostly about fixing those classes' compatibility issues. They were renaming and renumbering fields, which is now no longer allowed. We can be more accurate in future work.

File: server/tests/src/test/resources/proto/v1.proto (hunk 4 above)
Comment (tristantarrant) [reply server/tests/src/test/resources/proto/v1.proto:22]: I originally wanted to have these in the server testsuite, but it was easier to run/debug in the rest testsuite :). I'll remove them

**Comment by fax4ever** [comment]
Merged, thanks @tristantarrant

**Pull request merged**